// scoutos/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScoutClient = void 0;
const Client_1 = require("./api/resources/workflows/client/Client");
const Client_2 = require("./api/resources/environments/client/Client");
const Client_3 = require("./api/resources/revisions/client/Client");
const Client_4 = require("./api/resources/usage/client/Client");
const Client_5 = require("./api/resources/workflowLogs/client/Client");
const Client_6 = require("./api/resources/copilots/client/Client");
const Client_7 = require("./api/resources/collections/client/Client");
const Client_8 = require("./api/resources/tables/client/Client");
const Client_9 = require("./api/resources/documents/client/Client");
class ScoutClient {
    constructor(_options = {}) {
        this._options = _options;
    }
    get workflows() {
        var _a;
        return ((_a = this._workflows) !== null && _a !== void 0 ? _a : (this._workflows = new Client_1.Workflows(this._options)));
    }
    get environments() {
        var _a;
        return ((_a = this._environments) !== null && _a !== void 0 ? _a : (this._environments = new Client_2.Environments(this._options)));
    }
    get revisions() {
        var _a;
        return ((_a = this._revisions) !== null && _a !== void 0 ? _a : (this._revisions = new Client_3.Revisions(this._options)));
    }
    get usage() {
        var _a;
        return ((_a = this._usage) !== null && _a !== void 0 ? _a : (this._usage = new Client_4.Usage(this._options)));
    }
    get workflowLogs() {
        var _a;
        return ((_a = this._workflowLogs) !== null && _a !== void 0 ? _a : (this._workflowLogs = new Client_5.WorkflowLogs(this._options)));
    }
    get copilots() {
        var _a;
        return ((_a = this._copilots) !== null && _a !== void 0 ? _a : (this._copilots = new Client_6.Copilots(this._options)));
    }
    get collections() {
        var _a;
        return ((_a = this._collections) !== null && _a !== void 0 ? _a : (this._collections = new Client_7.Collections(this._options)));
    }
    get tables() {
        var _a;
        return ((_a = this._tables) !== null && _a !== void 0 ? _a : (this._tables = new Client_8.Tables(this._options)));
    }
    get documents() {
        var _a;
        return ((_a = this._documents) !== null && _a !== void 0 ? _a : (this._documents = new Client_9.Documents(this._options)));
    }
}
exports.ScoutClient = ScoutClient;
