// scoutos/serialization/types/BlockInput.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockInput = void 0;
const core = __importStar(require("../../core"));
const BlockInputBlockConfigItem_1 = require("./BlockInputBlockConfigItem");
const Dependency_1 = require("./Dependency");
const NodeUi_1 = require("./NodeUi");
exports.BlockInput = core.serialization.object({
    block_archetype_id: core.serialization.string(),
    block_config: core.serialization.list(BlockInputBlockConfigItem_1.BlockInputBlockConfigItem).optional(),
    block_display_name: core.serialization.string(),
    block_id: core.serialization.string(),
    block_inputs: core.serialization.record(core.serialization.string(), core.serialization.unknown()).optional(),
    block_is_output: core.serialization.boolean().optional(),
    dependencies: core.serialization.list(Dependency_1.Dependency).optional(),
    input_schema: core.serialization.record(core.serialization.string(), core.serialization.unknown()).optional(),
    output_schema: core.serialization.list(core.serialization.unknown()).optional(),
    ui: NodeUi_1.NodeUi.optional(),
    trigger_config: core.serialization.record(core.serialization.string(), core.serialization.unknown()).optional(),
    ignore_error: core.serialization.boolean().optional(),
});
