// scoutos/serialization/types/RunLog.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RunLog = void 0;
const core = __importStar(require("../../core"));
const BlockRun_1 = require("./BlockRun");
exports.RunLog = core.serialization.object({
    workflow_display_name: core.serialization.string(),
    workflow_run_id: core.serialization.string(),
    workflow_id: core.serialization.string(),
    status: core.serialization.string(),
    session_id: core.serialization.string(),
    cost: core.serialization.number(),
    blocks: core.serialization.record(core.serialization.string(), BlockRun_1.BlockRun),
    output: core.serialization.unknown(),
    elapsed_time_ms: core.serialization.number(),
    timestamp_start: core.serialization.date(),
    timestamp_end: core.serialization.date(),
    message: core.serialization.string().optional(),
    gcs_log_path: core.serialization.string().optional(),
});
