// scoutos/dist/api/resources/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.collections = exports.copilots = exports.usage = exports.revisions = exports.environments = exports.documents = exports.tables = exports.workflowLogs = exports.workflows = void 0;
exports.workflows = __importStar(require("./workflows"));
__exportStar(require("./workflows/types"), exports);
exports.workflowLogs = __importStar(require("./workflowLogs"));
__exportStar(require("./workflowLogs/types"), exports);
exports.tables = __importStar(require("./tables"));
__exportStar(require("./tables/types"), exports);
exports.documents = __importStar(require("./documents"));
__exportStar(require("./documents/types"), exports);
exports.environments = __importStar(require("./environments"));
exports.revisions = __importStar(require("./revisions"));
exports.usage = __importStar(require("./usage"));
exports.copilots = __importStar(require("./copilots"));
exports.collections = __importStar(require("./collections"));
__exportStar(require("./workflows/client/requests"), exports);
__exportStar(require("./environments/client/requests"), exports);
__exportStar(require("./usage/client/requests"), exports);
__exportStar(require("./workflowLogs/client/requests"), exports);
__exportStar(require("./copilots/client/requests"), exports);
__exportStar(require("./tables/client/requests"), exports);
