// scoutos/dist/serialization/types/BlockRunCompleted.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockRunCompleted = void 0;
const core = __importStar(require("../../core"));
const BlockRunCompletedEnvironment_1 = require("./BlockRunCompletedEnvironment");
const BlockRunCompletedData_1 = require("./BlockRunCompletedData");
const EventVersion_1 = require("./EventVersion");
exports.BlockRunCompleted = core.serialization.object({
    organization_id: core.serialization.string(),
    id: core.serialization.string().optional(),
    correlation_id: core.serialization.string().optional(),
    name: core.serialization.stringLiteral("block_run_completed"),
    environment: BlockRunCompletedEnvironment_1.BlockRunCompletedEnvironment,
    timestamp: core.serialization.string().optional(),
    data: BlockRunCompletedData_1.BlockRunCompletedData,
    price: core.serialization.string(),
    event_type: core.serialization.string(),
    version: EventVersion_1.EventVersion,
});
