// scoutos/dist/serialization/types/BlockConfigItemNumber.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockConfigItemNumber = void 0;
const core = __importStar(require("../../core"));
const BlockConfigItemNumberDefaultValue_1 = require("./BlockConfigItemNumberDefaultValue");
const BlockConfigItemNumberMinimumValue_1 = require("./BlockConfigItemNumberMinimumValue");
const BlockConfigItemNumberMaximumValue_1 = require("./BlockConfigItemNumberMaximumValue");
const BlockConfigItemNumberValue_1 = require("./BlockConfigItemNumberValue");
exports.BlockConfigItemNumber = core.serialization.object({
    id: core.serialization.string(),
    display_name: core.serialization.string(),
    type: core.serialization.stringLiteral("number"),
    default_value: BlockConfigItemNumberDefaultValue_1.BlockConfigItemNumberDefaultValue.optional(),
    minimum_value: BlockConfigItemNumberMinimumValue_1.BlockConfigItemNumberMinimumValue.optional(),
    maximum_value: BlockConfigItemNumberMaximumValue_1.BlockConfigItemNumberMaximumValue.optional(),
    value: BlockConfigItemNumberValue_1.BlockConfigItemNumberValue.optional(),
});
